flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java

```diff
@@ -0,0 +1,429 @@
+		} else {
+			return "TRUE";
+		}
+	}
+
+	// Predicate conversion for filter push-down.
+
+	private Predicate toOrcPredicate(Expression pred) {
+		if (pred instanceof Or) {
+			Predicate c1 = toOrcPredicate(((Or) pred).left());
+			Predicate c2 = toOrcPredicate(((Or) pred).right());
+			if (c1 == null || c2 == null) {
+				return null;
+			} else {
+				return new OrcRowInputFormat.Or(c1, c2);
+			}
+		} else if (pred instanceof Not) {
+			Predicate c = toOrcPredicate(((Not) pred).child());
+			if (c == null) {
+				return null;
+			} else {
+				return new OrcRowInputFormat.Not(c);
+			}
+		} else if (pred instanceof BinaryComparison) {
+
+			BinaryComparison binComp = (BinaryComparison) pred;
+
+			if (!isValid(binComp)) {
+				// not a valid predicate
+				return null;
+			}
+			PredicateLeaf.Type litType = getLiteralType(binComp);
+			if (litType == null) {
+				// unsupported literal type
+				return null;
+			}
+
+			boolean literalOnRight = literalOnRight(binComp);
+			String colName = getColumnName(binComp);
+			Serializable literal = (Serializable) getLiteral(binComp);
+
+			if (pred instanceof EqualTo) {
+				return new OrcRowInputFormat.Equals(colName, litType, literal);
+			} else if (pred instanceof NotEqualTo) {
+				return new OrcRowInputFormat.Not(
+					new OrcRowInputFormat.Equals(colName, litType, literal));
+			} else if (pred instanceof GreaterThan) {
+				if (literalOnRight) {
+					return new OrcRowInputFormat.Not(
+						new OrcRowInputFormat.LessThanEquals(colName, litType, literal));
+				} else {
+					return new OrcRowInputFormat.LessThan(colName, litType, literal);
+				}
+			} else if (pred instanceof GreaterThanOrEqual) {
+				if (literalOnRight) {
+					return new OrcRowInputFormat.Not(
+						new OrcRowInputFormat.LessThan(colName, litType, literal));
+				} else {
+					return new OrcRowInputFormat.LessThanEquals(colName, litType, literal);
+				}
+			} else if (pred instanceof LessThan) {
+				if (literalOnRight) {
+					return new OrcRowInputFormat.LessThan(colName, litType, literal);
+				} else {
+					return new OrcRowInputFormat.Not(
+						new OrcRowInputFormat.LessThanEquals(colName, litType, literal));
+				}
+			} else if (pred instanceof LessThanOrEqual) {
+				if (literalOnRight) {
+					return new OrcRowInputFormat.LessThanEquals(colName, litType, literal);
+				} else {
+					return new OrcRowInputFormat.Not(
+						new OrcRowInputFormat.LessThan(colName, litType, literal));
+				}
+			} else {
+				// unsupported predicate
+				return null;
+			}
+		} else if (pred instanceof UnaryExpression) {
+
+			UnaryExpression unary = (UnaryExpression) pred;
+			if (!isValid(unary)) {
+				// not a valid predicate
+				return null;
+			}
+			PredicateLeaf.Type colType = toOrcType(((UnaryExpression) pred).child().resultType());
+			if (colType == null) {
+				// unsupported type
+				return null;
+			}
+
+			String colName = getColumnName(unary);
+
+			if (pred instanceof IsNull) {
```

Comment:
I don't know how the ORC predicates are evaluated but if they comply with SQLs three-valued logic you cannot use this transformations. See: https://en.wikibooks.org/wiki/Structured_Query_Language/NULLs_and_the_Three_Valued_Logic#Step_2:_Boolean_Operations_within_3VL and `ScalarOperators`.